*OBJECTIVE*:

To focus on models to determine whether a patient admitted to an ICU has been diagnosed with a particular type of diabetes, Diabetes Mellitus. Using data from the first 24 hours of intensive care.

*Some Important features from the provided dataset*


---



* Age: Diabetes prevalence increases with age, so older individuals may have a higher risk.

* BMI (Body Mass Index): High BMI is associated with an increased risk of diabetes.

* Gender: Some studies suggest that men may have a slightly higher risk of diabetes compared to women.

* Ethnicity: Certain ethnic groups, such as African Americans, Hispanics, Native Americans, and Asian Americans, have a higher risk of diabetes.

* Glucose levels: Elevated glucose levels (glucose_apache) are indicative of diabetes.

* Albumin levels: Abnormal albumin levels (albumin_apache) can be associated with diabetes complications.

* Creatinine levels: Elevated creatinine levels (creatinine_apache) may indicate kidney damage, a complication of diabetes.

* Other relevant medical history: Conditions like immunosuppression, leukemia, lymphoma, solid tumors with metastasis, AIDS, cirrhosis, and hepatic failure can either predispose individuals to diabetes or complicate its management.

# Importing Data

## Initial Data Exploration: Loading and Displaying DataFrame



Importing the necessary libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import preprocessing
from sklearn.impute import KNNImputer
from IPython.display import display, HTML
import warnings
warnings.filterwarnings("ignore")

Reading the CSV file into a dataFrame and displaying the first few rows of the DataFrame to examine its structure and contents

In [2]:
df = pd.read_csv('TrainingWiDS2021.csv')
pd.set_option('display.max_columns', None)
df.head()

Unnamed: 0  encounter_id  hospital_id   age        bmi  elective_surgery  \
0           1        214826          118  68.0  22.732803                 0   
1           2        246060           81  77.0  27.421875                 0   
2           3        276985          118  25.0  31.952749                 0   
3           4        262220          118  81.0  22.635548                 1   
4           5        201746           33  19.0        NaN                 0   

   ethnicity gender  height hospital_admit_source           icu_admit_source  \
0  Caucasian      M   180.3                 Floor                      Floor   
1  Caucasian      F   160.0                 Floor                      Floor   
2  Caucasian      F   172.7  Emergency Department       Accident & Emergency   
3  Caucasian      F   165.1        Operating Room  Operating Room / Recovery   
4  Caucasian      M   188.0                   NaN       Accident & Emergency   

   icu_id icu_stay_type      icu_type  pre_icu_los_days  readmission_status  \
0      92         admit         CTICU          0.541667                   0   
1      90         admit  Med-Surg ICU          0.927778                   0   
2      93         admit  Med-Surg ICU          0.000694                   0   
3      92         admit         CTICU          0.000694                   0   
4      91         admit  Med-Surg ICU          0.073611                   0   

   weight  albumin_apache  apache_2_diagnosis  apache_3j_diagnosis  \
0    73.9             2.3               113.0               502.01   
1    70.2             NaN               108.0               203.01   
2    95.3             NaN               122.0               703.03   
3    61.7             NaN               203.0              1206.03   
4     NaN             NaN               119.0               601.01   

   apache_post_operative  arf_apache  bilirubin_apache  bun_apache  \
0                      0           0               0.4        31.0   
1                      0           0               NaN         9.0   
2                      0           0               NaN         NaN   
3                      1           0               NaN         NaN   
4                      0           0               NaN         NaN   

   creatinine_apache  fio2_apache  gcs_eyes_apache  gcs_motor_apache  \
0               2.51          NaN              3.0               6.0   
1               0.56          1.0              1.0               3.0   
2                NaN          NaN              3.0               6.0   
3                NaN          0.6              4.0               6.0   
4                NaN          NaN              NaN               NaN   

   gcs_unable_apache  gcs_verbal_apache  glucose_apache  heart_rate_apache  \
0                0.0                4.0           168.0              118.0   
1                0.0                1.0           145.0              120.0   
2                0.0                5.0             NaN              102.0   
3                0.0                5.0           185.0              114.0   
4                NaN                NaN             NaN               60.0   

   hematocrit_apache  intubated_apache  map_apache  paco2_apache  \
0               27.4                 0        40.0           NaN   
1               36.9                 0        46.0          37.0   
2                NaN                 0        68.0           NaN   
3               25.9                 1        60.0          30.0   
4                NaN                 0       103.0           NaN   

   paco2_for_ph_apache  pao2_apache  ph_apache  resprate_apache  \
0                  NaN          NaN        NaN             36.0   
1                 37.0         51.0       7.45             33.0   
2                  NaN          NaN        NaN             37.0   
3                 30.0        142.0       7.39              4.0   
4                  NaN          NaN        NaN             16.0   

   sodium_apache

# Understanding Dataset

Obtaining a detailed statistical summary and information about the DataFrame's structure

Getting Summary of the DataFrame

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130157 entries, 0 to 130156
Columns: 181 entries, Unnamed: 0 to diabetes_mellitus
dtypes: float64(157), int64(18), object(6)
memory usage: 179.7+ MB


The dataFrame consists of 130,157 rows and 181 columns. The data types of these columns include float64 datatype for 157 columns, int64 datatype for 18 columns, and object datatype (Categorical) for 6 columns.

Getting Statistics summary of Numeric columns

In [4]:
pd.set_option('display.max_rows', None)
df.describe().transpose()

count           mean           std  \
Unnamed: 0                   130157.0   65079.000000  37573.233831   
encounter_id                 130157.0  213000.856519  38109.828146   
hospital_id                  130157.0     106.102131     63.482277   
age                          125169.0      61.995103     16.822880   
bmi                          125667.0      29.110260      8.262776   
elective_surgery             130157.0       0.189840      0.392176   
height                       128080.0     169.607219     10.833085   
icu_id                       130157.0     662.428344    304.259843   
pre_icu_los_days             130157.0       0.839933      2.485337   
readmission_status           130157.0       0.000000      0.000000   
weight                       126694.0      83.791104     24.963063   
albumin_apache                51994.0       2.886149      0.689812   
apache_2_diagnosis           128472.0     185.492683     85.858208   
apache_3j_diagnosis          129292.0     565.994296    466.510850   
apache_post_operative        130157.0       0.207111      0.405238   
arf_apache                   130157.0       0.027997      0.164965   
bilirubin_apache              47597.0       1.201222      2.351994   
bun_apache                   104746.0      25.718070     20.690041   
creatinine_apache            105275.0       1.481629      1.543535   
fio2_apache                   30437.0       0.595735      0.262922   
gcs_eyes_apache              127967.0       3.488290      0.939831   
gcs_motor_apache             127967.0       5.484828      1.271039   
gcs_unable_apache            129448.0       0.011441      0.106349   
gcs_verbal_apache            127967.0       4.030203      1.538528   
glucose_apache               115461.0     160.141416     90.701327   
heart_rate_apache            129848.0      99.854530     30.759505   
hematocrit_apache            103399.0      32.975817      6.834576   
intubated_apache             130157.0       0.156626      0.363449   
map_apache                   129737.0      87.193046     41.908109   
paco2_apache                  30437.0      42.161246     12.267414   
paco2_for_ph_apache           30437.0      42.161246     12.267414   
pao2_apache                   30437.0     132.061737     84.958826   
ph_apache                     30437.0       7.352154      0.098423   
resprate_apache              129349.0      25.150603     15.024730   
sodium_apache                105638.0     137.945260      5.303840   
temp_apache                  123546.0      36.420638      0.857584   
urineoutput_apache            66990.0    1800.803417   1456.551481   
ventilated_apache            130157.0       0.330432      0.470370   
wbc_apache                   100682.0      12.187662      6.931023   
d1_diasbp_invasive_max        35089.0      79.261563     21.693320   
d1_diasbp_invasive_min        35089.0      46.969706     12.736433   
d1_diasbp_max                129880.0      88.575516     20.162800   
d1_diasbp_min                129880.0      49.862989     13.397152   
d1_diasbp_noninvasive_max    128521.0      88.729344     20.155701   
d1_diasbp_noninvasive_min    128521.0      49.917492     13.427746   
d1_heartrate_max             129895.0     103.296794     21.982940   
d1_heartrate_min             129895.0      70.408530     17.141875   
d1_mbp_invasive_max           35289.0     114.432769     49.182744   
d1_mbp_invasive_min           35289.0      62.211596     17.845442   
d1_mbp_max                   129830.0     104.228522     20.910213   
d1_mbp_min                   129830.0      64.353462     15.433092   
d1_mbp_noninvasive_max       127929.0     104.183969     20.802509   
d1_mbp_noninvasive_min       127929.0      64.384111     15.458592   
d1_resprate_max              129474.0      28.719388     10.567866   
d1_resprate_min              129474.0      12.690841      5.046519   
d1_spo2_max                  129625.0      99.276104      1.722287   
d1_spo2_min                  129625.0     

# Data Cleaning - Part 1

## Handling Rows & Columns

Removing leading and trailing white spaces in column names

In [5]:
df=df.rename(columns=lambda x: x.strip())
cols=df.columns
display(HTML(df.head().to_html()))

Summarizing the unique value information for each column in the DataFrame, to help us understand the diversity and cardinality of data within each column, aiding in data quality assessment, feature selection, and potential preprocessing steps.

In [6]:
column_unique_info = []
for col in df.columns:
    column_unique_info.append({
        'Column': col,
        'Data Type': df[col].dtype,
        'Unique Values Count': df[col].nunique(),
        'Unique Values' : df[col].unique()
    })
column_info_df = pd.DataFrame(column_unique_info)

display(HTML(column_info_df.sort_values(by=["Data Type", "Unique Values Count"]).to_html()))

We can observe that the 'readmission_status' column consistently holds a value of 0 throughout the dataset. Additionally, there are instances where individuals 'age' was recorded as 0, indicating potential data entry errors that require investigation and cleaning.

* Dropping 'Unnamed: 0', 'encounter_id' as they serve as unique identifiers, and are not useful for model training.
* Dropping 'readmission_status' since it contains only one value, '0', making it non-informative for analysis.

In [7]:
df.drop(['Unnamed: 0', 'encounter_id', 'readmission_status'], axis=1, inplace=True)
df.head()

hospital_id   age        bmi  elective_surgery  ethnicity gender  height  \
0          118  68.0  22.732803                 0  Caucasian      M   180.3   
1           81  77.0  27.421875                 0  Caucasian      F   160.0   
2          118  25.0  31.952749                 0  Caucasian      F   172.7   
3          118  81.0  22.635548                 1  Caucasian      F   165.1   
4           33  19.0        NaN                 0  Caucasian      M   188.0   

  hospital_admit_source           icu_admit_source  icu_id icu_stay_type  \
0                 Floor                      Floor      92         admit   
1                 Floor                      Floor      90         admit   
2  Emergency Department       Accident & Emergency      93         admit   
3        Operating Room  Operating Room / Recovery      92         admit   
4                   NaN       Accident & Emergency      91         admit   

       icu_type  pre_icu_los_days  weight  albumin_apache  apache_2_diagnosis  \
0         CTICU          0.541667    73.9             2.3               113.0   
1  Med-Surg ICU          0.927778    70.2             NaN               108.0   
2  Med-Surg ICU          0.000694    95.3             NaN               122.0   
3         CTICU          0.000694    61.7             NaN               203.0   
4  Med-Surg ICU          0.073611     NaN             NaN               119.0   

   apache_3j_diagnosis  apache_post_operative  arf_apache  bilirubin_apache  \
0               502.01                      0           0               0.4   
1               203.01                      0           0               NaN   
2               703.03                      0           0               NaN   
3              1206.03                      1           0               NaN   
4               601.01                      0           0               NaN   

   bun_apache  creatinine_apache  fio2_apache  gcs_eyes_apache  \
0        31.0               2.51          NaN              3.0   
1         9.0               0.56          1.0              1.0   
2         NaN                NaN          NaN              3.0   
3         NaN                NaN          0.6              4.0   
4         NaN                NaN          NaN              NaN   

   gcs_motor_apache  gcs_unable_apache  gcs_verbal_apache  glucose_apache  \
0               6.0                0.0                4.0           168.0   
1               3.0                0.0                1.0           145.0   
2               6.0                0.0                5.0             NaN   
3               6.0                0.0                5.0           185.0   
4               NaN                NaN                NaN             NaN   

   heart_rate_apache  hematocrit_apache  intubated_apache  map_apache  \
0              118.0               27.4                 0        40.0   
1              120.0               36.9                 0        46.0   
2              102.0                NaN                 0        68.0   
3              114.0               25.9                 1        60.0   
4               60.0                NaN                 0       103.0   

   paco2_apache  paco2_for_ph_apache  pao2_apache  ph_apache  resprate_apache  \
0           NaN                  NaN          NaN        NaN             36.0   
1          37.0                 37.0         51.0       7.45             33.0   
2           NaN                  NaN          NaN        NaN             37.0   
3          30.0                 30.0        142.0       7.39              4.0   
4           NaN                  NaN          NaN        NaN             16.0   

   sodium_apache  temp_apache  urineoutput_apache  ventilated_apache  \
0          134.0         39.3                 NaN                  0   
1          145.0         35.1                 NaN                  1   
2            NaN         36.7                 NaN                  0   
3            NaN         34.8      

Observing records with 'age' as 0

In [8]:
df[df['age'] == 0.0]

hospital_id  age        bmi  elective_surgery         ethnicity  \
10394            85  0.0        NaN                 0               NaN   
10953            85  0.0        NaN                 0               NaN   
12376           169  0.0        NaN                 0               NaN   
12538           177  0.0  34.561404                 0               NaN   
13298           177  0.0  26.162761                 0  African American   
14122           177  0.0  21.342855                 0               NaN   
25765           161  0.0        NaN                 0         Caucasian   
34590            70  0.0  16.700592                 0               NaN   
35300            70  0.0  42.657614                 0               NaN   
35429            70  0.0  34.068435                 0               NaN   
36314            70  0.0  34.765625                 0               NaN   
36526            70  0.0  32.218146                 0               NaN   
37477            16  0.0  30.511019                 0               NaN   
37748            16  0.0        NaN                 0               NaN   
38902            16  0.0  33.360112                 0               NaN   
39026           199  0.0        NaN                 0               NaN   
39261           199  0.0        NaN                 0               NaN   
89589           167  0.0        NaN                 0               NaN   
89608           167  0.0        NaN                 0               NaN   
89616           167  0.0        NaN                 0               NaN   
89626           167  0.0        NaN                 0               NaN   
89629           167  0.0        NaN                 0               NaN   
89637           167  0.0        NaN                 0               NaN   
89659           167  0.0        NaN                 0               NaN   
89662           167  0.0        NaN                 0               NaN   
89664           167  0.0        NaN                 0               NaN   
89665           167  0.0        NaN                 0               NaN   
89697           167  0.0        NaN                 0               NaN   
114792          141  0.0  36.881452                 0               NaN   
126760            7  0.0  16.110362                 0               NaN   

       gender  height hospital_admit_source           icu_admit_source  \
10394     NaN     NaN                   NaN                        NaN   
10953     NaN   165.1                   NaN                        NaN   
12376     NaN     NaN                   NaN                        NaN   
12538       M   170.1                   NaN       Accident & Emergency   
13298       F   157.5  Emergency Department       Accident & Emergency   
14122       F   193.0                   NaN                      Floor   
25765       M     NaN  Emergency Department       Accident & Emergency   
34590     NaN   162.5                   NaN  Operating Room / Recovery   
35300     NaN   182.9         Recovery Room  Operating Room / Recovery   
35429     NaN   177.8        Operating Room  Operating Room / Recovery   
36314     NaN   160.0        Operating Room  Operating Room / Recovery   
36526     NaN   167.6        Operating Room  Operating Room / Recovery   
37477     NaN   172.7                   NaN                      Floor   
37748     NaN     NaN  Emergency Department       Accident & Emergency   
38902     NaN   162.6  Emergency Department       Accident & Emergency   
39026     NaN     NaN                   NaN                        NaN   
39261     NaN     NaN                   NaN                        NaN   
89589     NaN     NaN                   NaN                        NaN   
89608     NaN     NaN                   NaN                        NaN   
89616     NaN     NaN                   NaN                        NaN   
89626     NaN     NaN  Emergency Department       Accident & Emergency   
89629     NaN     NaN                   NaN           

In [10]:
len(df[df['age'] == 0.0])

30

As observed from the above result, some observations have an age recorded as 0, and mostly these records also contain NaN values. Therefore, we are removing these records.

In [11]:
df = df[df['age'] != 0.0]

Verifying if there are any remaining records with an age of 0.

In [12]:
df[df['age'] == 0.0]

Empty DataFrame
Columns: [hospital_id, age, bmi, elective_surgery, ethnicity, gender, height, hospital_admit_source, icu_admit_source, icu_id, icu_stay_type, icu_type, pre_icu_los_days, weight, albumin_apache, apache_2_diagnosis, apache_3j_diagnosis, apache_post_operative, arf_apache, bilirubin_apache, bun_apache, creatinine_apache, fio2_apache, gcs_eyes_apache, gcs_motor_apache, gcs_unable_apache, gcs_verbal_apache, glucose_apache, heart_rate_apache, hematocrit_apache, intubated_apache, map_apache, paco2_apache, paco2_for_ph_apache, pao2_apache, ph_apache, resprate_apache, sodium_apache, temp_apache, urineoutput_apache, ventilated_apache, wbc_apache, d1_diasbp_invasive_max, d1_diasbp_invasive_min, d1_diasbp_max, d1_diasbp_min, d1_diasbp_noninvasive_max, d1_diasbp_noninvasive_min, d1_heartrate_max, d1_heartrate_min, d1_mbp_invasive_max, d1_mbp_invasive_min, d1_mbp_max, d1_mbp_min, d1_mbp_noninvasive_max, d1_mbp_noninvasive_min, d1_resprate_max, d1_resprate_min, d1_spo2_max, d1_spo2_min, d1_sysbp_invasive_max, d1_sysbp_invasive_min, d1_sysbp_max, d1_sysbp_min, d1_sysbp_noninvasive_max, d1_sysbp_noninvasive_min, d1_temp_max, d1_temp_min, h1_diasbp_invasive_max, h1_diasbp_invasive_min, h1_diasbp_max, h1_diasbp_min, h1_diasbp_noninvasive_max, h1_diasbp_noninvasive_min, h1_heartrate_max, h1_heartrate_min, h1_mbp_invasive_max, h1_mbp_invasive_min, h1_mbp_max, h1_mbp_min, h1_mbp_noninvasive_max, h1_mbp_noninvasive_min, h1_resprate_max, h1_resprate_min, h1_spo2_max, h1_spo2_min, h1_sysbp_invasive_max, h1_sysbp_invasive_min, h1_sysbp_max, h1_sysbp_min, h1_sysbp_noninvasive_max, h1_sysbp_noninvasive_min, h1_temp_max, h1_temp_min, d1_albumin_max, d1_albumin_min, d1_bilirubin_max, d1_bilirubin_min, d1_bun_max, d1_bun_min, ...]
Index: []

## Handling duplicates

### Checking for Duplicate Rows

In [13]:
# Check for duplicate records in the DataFrame
duplicate_rows = df[df.duplicated()]

# Display the duplicate rows, if any
if not duplicate_rows.empty:
    print("Duplicate records as:")
    print(duplicate_rows)
else:
    print("No duplicate records found.")

No duplicate records found.


We can observe that there are no duplicate rows in the dataset.

### Checking for Duplicate Columns

In [14]:
duplicate_cols = df.T.duplicated().T
duplicate_col_names = df.columns[duplicate_cols]
print(duplicate_col_names)

Index(['paco2_for_ph_apache', 'h1_inr_max', 'h1_inr_min'], dtype='object')


We can observe these duplicate columns in the dataset, namely 'paco2_for_ph_apache', 'h1_inr_max', and 'h1_inr_min'. Therefore, we are dropping these columns from the dataset.

In [15]:
df[duplicate_col_names].head()

paco2_for_ph_apache  h1_inr_max  h1_inr_min
0                  NaN         NaN         NaN
1                 37.0         1.3         1.3
2                  NaN         NaN         NaN
3                 30.0         1.6         1.1
4                  NaN         NaN         NaN

In [16]:
df.drop(duplicate_col_names, axis=1, inplace=True)

#Validating whether the columns have been successfully dropped
for col in duplicate_col_names:
    if col in df.columns:
        print(f"{col} is still present in the DataFrame.")
    else:
        print(f"{col} has been successfully dropped from the DataFrame.")

paco2_for_ph_apache has been successfully dropped from the DataFrame.
h1_inr_max has been successfully dropped from the DataFrame.
h1_inr_min has been successfully dropped from the DataFrame.


Downloading the Dataframe in CSV format after Initial Cleaning

In [17]:
df.to_csv('Initial_cleaned_df.csv', index=False)